src/window/concept_frame.rs

```diff
@@ -639,21 +641,40 @@ impl Frame for ConceptFrame {
-                                    .iter()
-                                    .find(|p| p.extension().map(|e| e == "ttf").unwrap_or(false))
-                                {
-                                    let mut font_data = Vec::new();
-                                    if let Ok(mut file) = ::std::fs::File::open(font) {
-                                        match file.read_to_end(&mut font_data) {
-                                            Ok(_) => self.font_data = Some(font_data),
-                                            Err(err) => {
-                                                log::error!("Could not read font file: {}", err)
+                                match fontconfig::FontConfig::new() {
+                                    Ok(font_config) => match font_config
+                                        .get_regular_family_fonts(&font_face)
+                                    {
+                                        Ok(regular_family_fonts) => match regular_family_fonts
+                                            .iter()
+                                            .find(|p| {
+                                                p.extension().map(|e| e == "ttf").unwrap_or(false)
+                                            }) {
+                                            Some(font) => {
+                                                let mut font_data = Vec::new();
+                                                match std::fs::File::open(font) {
+                                                    Ok(mut file) => match file
+                                                        .read_to_end(&mut font_data)
+                                                    {
+                                                        Ok(_) => self.font_data = Some(font_data),
+                                                        Err(err) => error!(
+                                                            "Could not read font file: {}",
+                                                            err
+                                                        ),
+                                                    },
+                                                    Err(err) => {
+                                                        error!("Could not open font file: {}", err)
+                                                    }
+                                                }
                                             }
+                                            None => error!("No ttf font found"),
+                                        },
+                                        Err(err) => {
+                                            error!("No regular sans font family found: {}", err)
                                         }
+                                    },
+                                    Err(_) => {
+                                        error!("Could not load font config file")
```

Comment:
You can probably clean this code up quite significantly by joining the results/options together instead of handling them one by one.

I don't think it's necessary to log different messages for every single error and it gets a bit messy.

Alternatively a procedural approach is usually much easier to read too, just early returning on error and assigning the result to a variable when everything went okay. That way it doesn't get so deeply nested and difficult to match up the match arms.

Reply:
What do you think about something like this?

```
if self.font_data.is_none() {
    match get_first_ttf_regular_sans_font(font_face) {
        Some(bytes) => self.font_data = bytes,
        None => error!("No font could be found"),
      }
  }
```
```
fn get_first_ttf_regular_sans_font(font_face: String) -> Option<Vec<u8>> {
        let font_config = fontconfig::FontConfig::new().ok()?;
        let regular_family_fonts = font_config.get_regular_family_fonts(&font_face).ok()?;
        let font = regular_family_fonts
            .iter()
            .find(|p| p.extension().map(|e| e == "ttf").unwrap_or(false))?;
        std::fs::read(font).ok()
    }
```

Reply:
I don't really see a huge benefit over just chaining things up yourself and matching the result, but it's an option.